db/blip_sync_context.go

```diff
@@ -616,11 +616,42 @@ func (bsc *BlipSyncContext) sendNoRev(sender *blip.Sender, docID, revID string,
 
 // Pushes a revision body to the client
 func (bsc *BlipSyncContext) sendRevision(sender *blip.Sender, docID, revID string, seq SequenceID, knownRevs map[string]bool, maxHistory int, handleChangesResponseCollection *DatabaseCollectionWithUser, collectionIdx *int) error {
-	rev, err := handleChangesResponseCollection.GetRev(bsc.loggingCtx, docID, revID, true, nil)
-	if base.IsDocNotFoundError(err) {
-		return bsc.sendNoRev(sender, docID, revID, collectionIdx, seq, err)
-	} else if err != nil {
-		return fmt.Errorf("failed to GetRev for doc %s with rev %s: %w", base.UD(docID).Redact(), base.MD(revID).Redact(), err)
+	rev, originalErr := handleChangesResponseCollection.GetRev(bsc.loggingCtx, docID, revID, true, nil)
+
+	// set if we find an alternative revision to send in the event the originally requested rev is unavailable
+	var replacedRevID string
+
+	collectionCtx, collectionErr := bsc.collections.get(collectionIdx)
+	if collectionErr != nil {
+		return collectionErr
+	}
+
+	if base.IsDocNotFoundError(originalErr) {
+		if !collectionCtx.sendReplacementRevs {
+			return bsc.sendNoRev(sender, docID, revID, collectionIdx, seq, originalErr)
+		}
+
+		base.DebugfCtx(bsc.loggingCtx, base.KeySync, "Unavailable revision for %q %s - finding replacement: %v", base.UD(docID), revID, originalErr)
+
+		// try the active rev instead as a replacement
+		replacementRev, replacementRevErr := handleChangesResponseCollection.GetRev(bsc.loggingCtx, docID, "", true, nil)
```

Comment:
There's some inefficiency here, as we would have fetched the current version of the doc from the bucket when the revCacheLoader attempted to fetch the original (not found) rev.  

Since the frequency of replacementRev is expected to be low I think we can live with this, but could consider filing a backlog enhancement to enhance revCacheLoader to have the ability to return activeRev during the initial fetch (or at least stick it in the rev cache).

Reply:
The stats do seem like they might be generally useful - both for understanding frequency and understanding what's going on in an environment.   I'll leave it to you whether to include in this PR or file a followup ticket.

Reply:
Will do as an immediate follow-up/PR (CBG-3960)